source/agora/consensus/EnrollmentManager.d

```diff
@@ -682,6 +638,76 @@ public class EnrollmentManager
         }
     }
 
+    /***************************************************************************
+
+        Generate and store pre-images for this cycle, as well as sparsely
+        selected preimages for other cycles.
+
+        This generates all pre-images needed for a cycle in the range of less
+        than two times of a validator cycle and store them in `cycle_preimages`.
+        Additionally, It generates and stores pre-images being used for defined
+        number of cycles in `preimage_rounds` when there is no pre-image needed.
+        This function is very expensive, but should be seldom called, and might
+        take more than 30ms for generating a hundred thousand number of them.
+
+        Params:
+            height = block height used to determine which range of
+                pre-images will be generated.
+
+    ***************************************************************************/
+
+    private void generatePreimages (ulong height) @safe nothrow
+    {
+        // This determines which range of preimages must be generated.
+        // In order to get hash values more than one cycle, the `start_height`
+        // is to be the height of the last preimage of next cycle. The value of
+        // `height / ValidatorSet.ValidatorCycle` is the index of previous
+        // cycle, so we need to plus 2 to the value in order to get the index
+        // of the next cycle.
+        ulong start_height =
+            ((height / ValidatorSet.ValidatorCycle) + 2) *
+                ValidatorSet.ValidatorCycle;
+
+        // Clear if recreating pre-images is needed
+        () @trusted {
```

Comment:
Can `@trusted` be applied only on `clear` or does something in the `if` make it `@system` ?